File: src/scope.ts
// Per-agent scopes (DESIGN.md § Scopes and context). Policy is fixed at
// `open()`, identity travels per call. One resolver lives here and every
// enforcement point — search, get, list, the write gate — asks it, so a
// namespace prefix cannot come to mean two different things in two files.
//
// Scopes are advisory containment at the library API, not security: any process
// with filesystem access can read or edit the notes directly. That is the
// files-are-truth contract, not a hole in it.
import type { VaultContext } from "./gate";
import { safe } from "./term";

/**
 * A namespace subtree and what it grants. `prefix` matches on **segment
 * boundaries only** (`ledger/` matches `ledger/q3.md`, never
 * `ledger-archive/q3.md`); `""` is the root rule and matches every note. A
 * permission left undefined is not "denied" — it defers to the next-shorter
 * matching rule.
 */
export type ScopeRule = { prefix: string; read?: boolean; write?: boolean };

/** agent name → its rules. Absent from `VaultOptions` means allow-all. */
export type ScopePolicy = Record<string, ScopeRule[]>;

export type Permission = "read" | "write";

/** A validated policy: agent → normalized rules, longest prefix first. */
export type CompiledPolicy = Map<string, ScopeRule[]>;

/**
 * What one call may touch, resolved once from the policy and the caller's
 * context — which it carries, because the write gate needs the same object for
 * provenance.
 */
export type Scope = {
  readonly ctx?: VaultContext;
  may(permission: Permission, path: string): boolean;
  /**
   * The read check as a SQL boolean over `n.path`, for the two filters that
   * have to run inside the query (search's over-fetched set and its link
   * expansion). Generated from the same normalized rules as `may`, so the two
   * cannot drift — `test/scope.test.ts` holds them to the same answers.
   */
  readonly readSql: { sql: string; params: (string | number)[] };
};

/** No policy: everything allowed, and the query filters compile away to `1`. */
export const ALLOW_ALL: Scope = {
  may: () => true,
  readSql: { sql: "1", params: [] },
};

/**
 * `ledger`, `/ledger/` and `ledger/` all name the subtree `ledger/`; `""`, `/`
 * and undefined name the vault root. Slashes at either end are decoration —
 * stored paths carry neither — and the trailing one is what makes a prefix
 * match whole segments instead of half a namespace name.
 */
export function normalizePrefix(prefix?: string): string {
  const ns = prefix?.replace(/^\/+|\/+$/g, "") ?? "";
  return ns === "" ? "" : `${ns}/`;
}

/**
 * Validate and normalize a policy at `open()`, or null when there is none.
 * Two rules that answer the same question two ways are refused rather than
 * resolved by array order, and so is a subtree an agent could write but not
 * read — a write-blind agent never sees its own notes as `similar`, so every
 * propose lands as `create`: a duplicate factory, not a scope.
 */
export function compileScopes(policy?: ScopePolicy): CompiledPolicy | null {
  if (policy === undefined) return null;
  const compiled: CompiledPolicy = new Map();
  for (const [agent, given] of Object.entries(policy)) {
    // Annotated on the variable, not the arrow: that is what lets TypeScript
    // read a call to it as "control flow ends here" and narrow what follows.
    const bad: (why: string) => never = (why) => {
      throw new Error(`vault: scope policy for ${JSON.stringify(safe(agent))}: ${why}`);
    };
    // A policy is operator configuration — it arrives from a config file, an
    // orchestrator, another process's JSON — so it is checked, not trusted to
    // be what its type says. `read: "false"` is *truthy*: unchecked, a policy
    // that meant to deny would grant, which is the one direction an allowlist
    // must never fail in.
    if (!Array.isArray(given)) bad("must be a list of rules");
    for (const rule of given as unknown[]) {
      if (rule === null || typeof rule !== "object") {
        bad(`a rule must be {prefix, read?, write?}, got ${JSON.stringify(rule)?.slice(0, 60)}`);
      }
      const r = rule as Partial<ScopeRule>;
      if (typeof r.prefix !== "string") bad("a rule's prefix must be a string");
      for (const permission of PERMISSIONS) {
        if (r[permission] !== undefined && typeof r[permission] !== "boolean") {
          bad(
            `${permission} must be true, false or absent, in rule ` +
              JSON.stringify(safe(r.prefix)),
          );
        }
      }
      // Stored paths are canonical, so a prefix that is not matches nothing —
      // and a deny rule that matches nothing is a deny that never fires.
      // Refused rather than normalized: `ledger/../x` is not a typo we get to
      // guess the intent of.
      const segments = r.prefix.replace(/^\/+|\/+$/g, "").split("/");
      const canonical = segments.every(
        (s, i) => s !== "." && s !== ".." && (s !== "" || i === 0),
      );
      if (r.prefix !== "" && !canonical) {
        bad(`${JSON.stringify(safe(r.prefix))} is not a canonical namespace`);
      }
    }
    const rules = given
      .map((r) => ({ ...r, prefix: normalizePrefix(r.prefix) }))
      // Longest first: resolution is "longest matching prefix that specifies
      // this permission wins", which is a first-match scan over this order.
      .sort((a, b) => b.prefix.length - a.prefix.length);

    const seen = new Map<string, ScopeRule>();
    for (const rule of rules) {
      const prior = seen.get(rule.prefix);
      for (const permission of PERMISSIONS) {
        if (rule[permission] !== undefined && prior?.[permission] !== undefined) {
          bad(`two rules for ${JSON.stringify(safe(rule.prefix))} both specify ${permission}`);
        }
      }
      seen.set(rule.prefix, {
        prefix: rule.prefix,
        read: rule.read ?? prior?.read,
        write: rule.write ?? prior?.write,
      });
    }
    // Resolution only changes at a rule's own prefix, so checking each one
    // covers every path under it.
    for (const { prefix } of rules) {
      if (resolve(rules, "write", prefix) && !resolve(rules, "read", prefix)) {
        bad(
          `${JSON.stringify(safe(prefix))} is writable but not readable — an agent that cannot ` +
            `see its own notes re-creates them on every propose`,
        );
      }
    }
    compiled.set(agent, rules);
  }
  return compiled;
}

/**
 * The scope in force for one call. A policy fails closed: a call with no
 * `VaultContext`, or one naming an agent the policy never mentions, throws
 * rather than returning a silently empty result — silence is how an
 * orchestrator typo makes an agent re-create the memory it thinks it lost.
 */
export function scopeFor(policy: CompiledPolicy | null, ctx?: VaultContext): Scope {
  if (policy === null) return { ...ALLOW_ALL, ctx };
  if (ctx === undefined) {
    throw new Error(
      "vault: this vault has a scope policy, so every call needs a VaultContext — {agent: '<name>'}",
    );
  }
  const rules = policy.get(ctx.agent);
  if (rules === undefined) {
    throw new Error(`vault: agent ${JSON.stringify(safe(ctx.agent))} has no scope in this vault`);
  }
  return {
    ctx,
    may: (permission, path) => resolve(rules, permission, path),
    readSql: readSql(rules),
  };
}

const PERMISSIONS: readonly Permission[] = ["read", "write"];

/** Longest matching prefix that specifies this permission decides; none ⇒ denied. */
function resolve(rules: ScopeRule[], permission: Permission, path: string): boolean {
  for (const rule of rules) {
    if (!path.startsWith(rule.prefix)) continue;
    const allowed = rule[permission];
    if (allowed !== undefined) return allowed;
  }
  return false;
}

/**
 * `resolve(rules, "read", n.path)` as SQL: a CASE over the rules in the same
 * longest-first order, so the first prefix that matches and specifies `read`
 * decides. `substr` rather than `like`, whose `%` and `_` would have to be
 * escaped out of a namespace name — and SQLite measures the prefix itself
 * (`length(?)`), because it counts characters where JS counts UTF-16 code
 * units: a JS length for `🔒secret/` is one too many, the substr never equals
 * the prefix, and a deny rule quietly stops denying.
 */
function readSql(rules: ScopeRule[]): Scope["readSql"] {
  const params: (string | number)[] = [];
  const whens = rules
    .filter((r) => r.read !== undefined)
    .map((r) => {
      params.push(r.prefix, r.prefix);
      return `when substr(n.path, 1, length(?)) = ? then ${r.read ? 1 : 0}`;
    });
  // No rule mentions `read` at all, so nothing is readable.
  if (whens.length === 0) return { sql: "0", params: [] };
  return { sql: `case ${whens.join(" ")} else 0 end`, params };
}

File: src/wilcus_vault/scope.py
"""Per-agent namespace scopes. Policy is fixed at open(), identity travels per call.

Scopes are advisory containment at the library API, not security: any process
with filesystem access can read or edit the notes directly.
"""

import json
from collections.abc import Mapping
from dataclasses import dataclass
from typing import Any, Literal, NotRequired, TypedDict

from .term import VaultError, safe

Permission = Literal["read", "write"]
PERMISSIONS: tuple[Permission, ...] = ("read", "write")


class ScopeRule(TypedDict):
    """A namespace subtree and what it grants. A permission left out defers to
    the next-shorter matching rule; `""` is the root rule."""

    prefix: str
    read: NotRequired[bool]
    write: NotRequired[bool]


ScopePolicy = Mapping[str, list[ScopeRule]]  # agent name -> its rules


@dataclass(frozen=True)
class VaultContext:
    """Who is calling, per call. `source` records what prompted the call."""

    agent: str
    source: str | None = None


@dataclass(frozen=True)
class Rule:
    prefix: str  # normalized: `ledger/`, or `""` for the root
    read: bool | None
    write: bool | None

    def get(self, permission: Permission) -> bool | None:
        return self.read if permission == "read" else self.write


CompiledPolicy = dict[str, list[Rule]]  # longest prefix first


def normalize_prefix(prefix: str | None) -> str:
    """`ledger`, `/ledger/` and `ledger/` all name `ledger/`; `""`, `/` and None name the root."""
    ns = (prefix or "").strip("/")
    return f"{ns}/" if ns else ""


def _resolve(rules: list[Rule], permission: Permission, path: str) -> bool:
    """Longest matching prefix that specifies this permission decides; none means denied."""
    for rule in rules:
        if path.startswith(rule.prefix):
            allowed = rule.get(permission)
            if allowed is not None:
                return allowed
    return False


@dataclass(frozen=True)
class Scope:
    """What one call may touch. `rules` None means no policy: everything allowed."""

    ctx: VaultContext | None = None
    rules: list[Rule] | None = None

    def may(self, permission: Permission, path: str) -> bool:
        return True if self.rules is None else _resolve(self.rules, permission, path)

    @property
    def read_sql(self) -> tuple[str, list[str]]:
        """The read check as a SQL boolean over `n.path`, for filters inside a query.

        A CASE over the rules in longest-first order. `substr` rather than
        `like`, whose wildcards would need escaping; SQLite measures the prefix
        itself so multi-byte characters count the same on both sides.
        """
        if self.rules is None:
            return "1", []
        params: list[str] = []
        whens = []
        for rule in self.rules:
            if rule.read is not None:
                params += [rule.prefix, rule.prefix]
                whens.append(f"when substr(n.path, 1, length(?)) = ? then {int(rule.read)}")
        if not whens:
            return "0", []
        return f"case {' '.join(whens)} else 0 end", params


ALLOW_ALL = Scope()


def compile_scopes(policy: ScopePolicy | None) -> CompiledPolicy | None:
    """Validate and normalize a policy at open(), or None when there is none.

    A policy is operator config, so it is checked rather than trusted: two
    rules answering one question two ways are refused, and so is a subtree an
    agent could write but not read (it would never see its own notes).
    """
    if policy is None:
        return None
    compiled: CompiledPolicy = {}
    for agent, given in policy.items():
        compiled[agent] = _compile_rules(agent, given)
    return compiled


def _compile_rules(agent: str, given: Any) -> list[Rule]:
    def bad(why: str) -> VaultError:
        return VaultError(f"vault: scope policy for {json.dumps(safe(agent))}: {why}")

    if not isinstance(given, list):
        raise bad("must be a list of rules")
    for raw in given:
        if not isinstance(raw, dict):
            raise bad(f"a rule must be {{prefix, read?, write?}}, got {json.dumps(raw)[:60]}")
        prefix = raw.get("prefix")
        if not isinstance(prefix, str):
            raise bad("a rule's prefix must be a string")
        for permission in PERMISSIONS:
            if permission in raw and not isinstance(raw[permission], bool):
                raise bad(
                    f"{permission} must be true, false or absent, "
                    f"in rule {json.dumps(safe(prefix))}"
                )
        # Stored paths are canonical, so a prefix that is not matches nothing,
        # and a deny that matches nothing never fires. Refused, not guessed at.
        segments = prefix.strip("/").split("/")
        canonical = all(
            s not in (".", "..") and (s != "" or i == 0) for i, s in enumerate(segments)
        )
        if prefix != "" and not canonical:
            raise bad(f"{json.dumps(safe(prefix))} is not a canonical namespace")

    rules = sorted(
        (Rule(normalize_prefix(r["prefix"]), r.get("read"), r.get("write")) for r in given),
        key=lambda r: len(r.prefix),
        reverse=True,
    )
    seen: dict[str, Rule] = {}
    for rule in rules:
        prior = seen.get(rule.prefix)
        for permission in PERMISSIONS:
            if rule.get(permission) is not None and prior and prior.get(permission) is not None:
                raise bad(
                    f"two rules for {json.dumps(safe(rule.prefix))} both specify {permission}"
                )
        seen[rule.prefix] = Rule(
            rule.prefix,
            rule.read if rule.read is not None else (prior.read if prior else None),
            rule.write if rule.write is not None else (prior.write if prior else None),
        )
    # Resolution only changes at a rule's own prefix, so checking each covers every path.
    for rule in rules:
        if _resolve(rules, "write", rule.prefix) and not _resolve(rules, "read", rule.prefix):
            raise bad(
                f"{json.dumps(safe(rule.prefix))} is writable but not readable — an agent that "
                "cannot see its own notes re-creates them on every propose"
            )
    return rules


def scope_for(policy: CompiledPolicy | None, ctx: VaultContext | None) -> Scope:
    """The scope in force for one call. A policy fails closed: no context, or an
    agent it never names, raises rather than answering silently empty."""
    if policy is None:
        return Scope(ctx=ctx)
    if ctx is None:
        raise VaultError(
            "vault: this vault has a scope policy, so every call needs a VaultContext — "
            "VaultContext(agent='<name>')"
        )
    rules = policy.get(ctx.agent)
    if rules is None:
        raise VaultError(f"vault: agent {json.dumps(safe(ctx.agent))} has no scope in this vault")
    return Scope(ctx=ctx, rules=rules)
